### PR TITLE
Added StudyConsentController

### PR DESCRIPTION
**Pull request by dr-willikers**

### CODE REVIEW

File: app/controllers/StudyConsentController.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        String studyKey = studyControllerService.getStudyByHostname(request()).getKey();
+        StudyConsent studyConsent = studyConsentDao.getConsent(studyKey);
+        if (studyConsent != null) {
+            studyConsentDao.setActive(studyConsent, false);
+        }
+        
+        JsonNode json = requestToJSON(request());
+        fieldsValid(json);
+        String path = json.get("path").asText();
+        int minAge = json.get("minAge").asInt();
```
Comment (alxdarksage) [inline app/controllers/StudyConsentController.java]: studyconsent fromJSON method, pass this into "add consent"

File: app/controllers/StudyConsentController.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            studyConsentDao.setActive(studyConsent, false);
+        }
+        
+        studyConsent = studyConsentDao.getConsent(studyKey, timestamp);
+        studyConsentDao.setActive(studyConsent, true);
+
+        return okResult("Consent document set as active.");
+    }
+
+    public Result deleteConsent(long timestamp) throws Exception {
```
Comment (alxdarksage) [inline app/controllers/StudyConsentController.java]: no delete in controller, add delete in service

File: test/org/sagebionetworks/bridge/TestUserAdminHelper.java [hunk 3]
```diff
@@ -26,61 +29,108 @@
...
         study = studyControllerService.getStudyByHostname("pd.sagebridge.org");
         SignIn admin = new SignIn(bridgeConfig.getProperty("admin.email"), bridgeConfig.getProperty("admin.password"));
         adminSession = authService.signIn(study, admin);
-        
-        userSession = userAdminService.createUser(adminSession.getUser(), testUser.getSignUp(), study, true, true);
+
+        UserSession userSession = userAdminService.createUser(adminSession.getUser(), testUser.getSignUp(), study,
+                true, true);
+        userSessionToken = userSession.getSessionToken();
+        cache.setUserSession(userSessionToken, userSession);
```
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/TestUserAdminHelper.java]: don't need cachel

File: test/org/sagebionetworks/bridge/TestUserAdminHelper.java [hunk 4]
```diff
@@ -26,61 +29,108 @@
...
     public TestUser getTestUser() {
         return testUser;
     }
 
+    public void addRoleToUser(String role) {
+        User user = cache.getUserSession(userSessionToken).getUser();
+
+        role = role.toLowerCase();
+        Set<String> roles = user.getRoles();
+        if (!roles.contains(role)) {
```
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/TestUserAdminHelper.java]: remove

Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/TestUserAdminHelper.java]: roles.add remove

File: test/org/sagebionetworks/bridge/dynamodb/DynamoStudyConsentDaoTest.java [hunk 5]
```diff
@@ -42,48 +43,55 @@ public void test() {
...
+//        consent = all.get(2);
+//        assertTrue(consent.getActive());
+//        assertEquals(consent1.getStudyKey(), consent.getStudyKey());
+//        assertEquals(consent1.getPath(), consent.getPath());
+//        assertEquals(consent1.getMinAge(), consent.getMinAge());
+//        // Delete all consents
+//        for (StudyConsent x : all) {
+//            studyConsentDao.deleteConsent(x.getStudyKey(), x.getCreatedOn());
+//        }
+//        all = studyConsentDao.getConsents("fake-study");
```
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoStudyConsentDaoTest.java]: re enable test

File: test/org/sagebionetworks/bridge/TestUserAdminHelper.java [hunk 6]
```diff
@@ -1,6 +1,9 @@
 package org.sagebionetworks.bridge;
```
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/TestUserAdminHelper.java:1]: add create admin user function

File: app/org/sagebionetworks/bridge/services/StudyConsentServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        this.studyConsentDao = studyConsentDao;
+    }
+
+    @Override
+    public StudyConsent addConsent(User caller, String studyKey, StudyConsentForm form) {
+        assertAdminUser(caller);
+
+        StudyConsent studyConsent = studyConsentDao.getConsent(studyKey);
+        if (studyConsent != null) {
+            studyConsentDao.setActive(studyConsent, false);
```
Comment (eric-wu) [inline app/org/sagebionetworks/bridge/services/StudyConsentServiceImpl.java]: Not sure we need to set the current one to inactive when adding a new consent (or rather new version).  The method is generating too much  side-effect.  Imagine that the method sets the current to inactive and then fails to add the new one.  There would be no active consent.  That said, I prefer single-policy methods, methods that do one thing only.

We can have multiple active versions. Dynamo set up the version (the timestamp) as the range key.  The method retrieving the active consent always retrieves the newest version.

File: app/org/sagebionetworks/bridge/services/StudyConsentServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        return studyConsentDao.getConsent(studyKey, timestamp);
+    }
+
+    @Override
+    public void activateConsent(User caller, String studyKey, long timestamp) {
+        assertAdminUser(caller);
+
+        StudyConsent studyConsent = studyConsentDao.getConsent(studyKey);
+        if (studyConsent != null) {
+            studyConsentDao.setActive(studyConsent, false);
```
Comment (eric-wu) [inline app/org/sagebionetworks/bridge/services/StudyConsentServiceImpl.java]: The same comment as the one for the addConsent() method.

**Pull request merged**